recipe_system/cal_service/remotedb.py

```diff
@@ -0,0 +1,191 @@
+# Defines the RemoteDB class for calibration returns. This is a high-level
+# interface to FITSstore. It may be subclassed in future
+
+from os import path, makedirs
+from io import BytesIO
+from pprint  import pformat
+from xml.dom import minidom
+
+import urllib.request
+import urllib.parse
+import urllib.error
+
+from .caldb import CalDB, CalReturn
+from .calrequestlib import get_cal_requests, generate_md5_digest, get_request
+from .calrequestlib import GetterError
+
+UPLOADCOOKIE = "qap_upload_processed_cal_ok"
+
+RESPONSESTR = """########## Request Data BEGIN ##########
+%(sequence)s
+########## Request Data END ##########
+
+########## Calibration Server Response BEGIN ##########
+%(response)s
+########## Calibration Server Response END ##########
+
+########## Nones Report (descriptors that returned None):
+%(nones)s
+########## Note: all descriptors shown above, scroll up.
+        """
+
+
+class RemoteDB(CalDB):
+    """
+    The class for remote calibration databases. It inherits from CalDB, but
+    also has the following attributes:
+
+    Attributes
+    ----------
+    server : str
+        URL of the server
+    store_science : bool
+        whether processed science images should be uploaded
+    _upload_cookie : str
+        the cookie to send when uploading files
+    _calmgr : str
+        the URL for making requests to the remote calibration manager
+    _proccal_url, _science_url : str
+        the URLs for uploading processed calibrations and processed science
+        images, respectively.
+    """
+    def __init__(self, server, name=None, valid_caltypes=None, get_cal=True,
+                 store_cal=False, store_science=False, procmode=None, log=None,
+                 upload_cookie=None):
+        if name is None:
+            name = server
+        super().__init__(name=name, get_cal=get_cal, store_cal=store_cal,
+                         log=log, valid_caltypes=valid_caltypes,
+                         procmode=procmode)
+        self.store_science = store_science
+        if not server.startswith("http"):  # allow https://
+            server = f"http://{server}"
+        self.server = server
+        self._calmgr = f"{self.server}/calmgr"
+        self._proccal_url = f"{self.server}/upload_processed_cal"
+        self._science_url = f"{self.server}/upload_file"
+        self._upload_cookie = upload_cookie or UPLOADCOOKIE
+
+    def _get_calibrations(self, adinputs, caltype=None, procmode=None,
```

Comment:
Does this have a particular reason to be private? Could we have some explanation in the docstring? (Same for the method below).

Reply:
The structure of the databases is that each one has a link to the next one in the list/cascade (whatever you want to call it). The private `_get_calibrations()` method only looks at its own database. The public `get_calibrations()` method (defined in the parent `CalDB` class) gets calibrations from its own database and then send any input files which didn't get a calibration on to the next database, so that's the one that should be used. It's the same for `_store_calibration()`.